Ad5770r iio driver

mirceacaprioru opened the pull request:
This patch adds the output filter resistor functionality. This is done using IIO_ENUM and IIO_ENUM_AVAILABLE. A new attribute has also been documented.

lclausen-adi commented:
How about using the standard LOW_PASS_FILTER_3DB_FREQUENCY property. Table 11 in the datasheet lists the corresponding cutoff frequencies for each resistor setting.

stefpopa commented:
Small typo in the commit message: resister -> resistor

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -453,6 +517,7 @@ static int ad5770r_init(struct ad5770r_state *st)
 	if (ret)
 		return ret;
 
+
```
nitpick: extra whitespace

stefpopa replied:
same as above

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -496,8 +561,8 @@ static int ad5770r_probe(struct spi_device *spi)
-			dev_err(&spi->dev, "Failed to enable vref regulators: %d\n",
-				ret);
+			dev_err(&spi->dev,
+				"Failed to enable vref regulators: %d\n", ret);
```
generally: if you want to do whitespace adjustments, they would typically go into their own commit;
otherwise, if we can live without them, then let's leave them as is;
whitespace changes should ideally have some added value [when doing them]

stefpopa replied:
different commit as well

stefpopa commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -398,6 +461,7 @@ static int ad5770r_channel_config(struct ad5770r_state *st)
 		ret = fwnode_property_read_u32(child, "num", &num);
 		if (ret)
 			return ret;
+
```
white space fixes belong in a different commit

commodo commented on Documentation/ABI/testing/sysfs-bus-iio-dac-ad5770r:
```diff
@@ -0,0 +1,20 @@
+What:		/sys/bus/iio/devices/iio:deviceX/out_currentY_filter_mode
```
the title needs to be a bit more descriptive ; something like `Documentation: iio: testing: add sysfs-bus-iio-dac-ad5770r`

mirceacaprioru commented:
Changelog V1->V2
- used LOW_PASS_FILTER_3DB_FREQUENCY property
- removed extra lines
- removed sysfs documentation

lclausen-adi commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -496,8 +570,6 @@ static int ad5770r_probe(struct spi_device *spi)
 		ret = regulator_enable(st->vref_reg);
 		if (ret)
-			dev_err(&spi->dev, "Failed to enable vref regulators: %d\n",
-				ret);
```
This looks like it was added by accident to the patch.

lclausen-adi commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -224,6 +244,55 @@ static int ad5770r_get_range(struct ad5770r_state *st,
+	regval = AD5770R_FILTER_RESISTOR_60_OHM;
+	best_freq = 0;
+	best_diff = 262000;
+	for (i = 0; i < ARRAY_SIZE(ad5770r_filter_freqs); i++)
```
It might make sense to iterate in the reverse order and then stop once ad5770r_filter_freqs[i] >= freq. You know that this will be the closest solution since the entries in the array are sorted. Then you don't have to worry about best_diff and all that stuff.

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -224,6 +244,55 @@ static int ad5770r_get_range(struct ad5770r_state *st,
+	if (regval >= AD5770R_FILTER_RESISTOR_5_6_KOHM)
+		*freq = ad5770r_filter_freqs[regval - AD5770R_FILTER_OFFSET];
+	else
+		*freq = ad5770r_filter_freqs[regval];
```
here you could do:
```
	if (regval >= AD5770R_FILTER_RESISTOR_5_6_KOHM)
		regval = regval - AD5770R_FILTER_RESISTOR_5_6_KOHM + 1;

	*freq = ad5770r_filter_freqs[regval];
```

commodo replied:
the idea is that you could get rid of `AD5770R_FILTER_OFFSET` ; it's one less definition to have

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -224,6 +244,55 @@ static int ad5770r_get_range(struct ad5770r_state *st,
+	if (best_freq > 0)
+		regval = best_freq + AD5770R_FILTER_OFFSET;
+	else
+		regval = best_freq;
```
here, you could do
```
	if (regval >= AD5770R_FILTER_RESISTOR_5_6_KOHM)
		regval = best_freq + (AD5770R_FILTER_RESISTOR_5_6_KOHM - 1);
	else
		regval = best_freq;
```

commodo commented:
rest looks good

mirceacaprioru commented:
Changelog V2 -> V3
- added available property for frequency
- removed offset define
- simplified set_freq function

lclausen-adi commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -367,15 +446,18 @@ static const struct iio_chan_spec_ext_info ad5770r_ext_info[] = {
+	.info_mask_separate = BIT(IIO_CHAN_INFO_RAW) |			\
+		BIT(IIO_CHAN_INFO_SCALE) |				\
+		BIT(IIO_CHAN_INFO_LOW_PASS_FILTER_3DB_FREQUENCY),	\
+	.info_mask_shared_by_all_available =				\
```
I'd make that shared_by_type_available. While there are no other types this only applies to the current channels and using shared_by_type_available gives the property a nicer name. This will allow libiio to associate the property with the channel instead of seeing it as a device property.

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -224,6 +243,45 @@ static int ad5770r_get_range(struct ad5770r_state *st,
+		if (ad5770r_filter_freqs[i] >= freq)
+			break;
+
+	if (i < AD5770R_FILTER_5_6_KOHM)
```
i'm having a bit of trouble following this logic [ may be the coffee that's not working ];
maybe it would be an idea to add the `AD5770R_FILTER_xxxx` in a static array and matching the order of the frequencies ;

anyone other thoughts [to make this easier to follow] ?

mirceacaprioru replied:
i could have another array with the frequencies in descending order to correspond with the register value. That would simplify the logic. This will duplicate the frequency array. Is this frown upon in kernel drivers ?

commodo replied:
my sentiment is that the added value [to code readability] would justify the extra static array;

lclausen-adi replied:
maybe a set of macros to translate between the two, so that the code that does the translation is out of line.

stefpopa commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -142,6 +152,15 @@ static struct ad5770r_output_modes ad5770r_rng_tbl[] = {
+	1400,
+	2800,
+	262000,
+};
```
nitpick: please check if you can fit all these values in one line

mirceacaprioru commented:
V3 -> V4
- change available property to shared_by_type_available
- added reverse array in order to simplify the logic in set and get frequency functions

mirceacaprioru commented:
v4 -> v5 (sorry for misinterpreting the reverse array)
- removed duplicate array
-added an array with enum register values thus the corespondence between frequencies and register values is one to one and the index is just passed.

commodo commented on drivers/iio/dac/ad5770r.c:
```diff
@@ -224,6 +247,49 @@ static int ad5770r_get_range(struct ad5770r_state *st,
+	if (ret < 0)
+		return ret;
+
+	for (i = 0; i <= ARRAY_SIZE(ad5770r_filter_reg_vals); i++)
```
`i < ARRAY_SIZE(ad5770r_filter_reg_vals)` ?
